### PR TITLE
Add include_loss_for_metrics

### LINKED ISSUE #32307: Make evaluation loss accessible inside compute_metrics()  in Trainer
### Feature request

Hi, I am requesting the feature to make evaluation loss accessible inside `compute_metrics()`  within the Trainer class, this will enable users to log loss dependent metrics during training, in my case I want to track perplexity.  By making the pre-computed validation loss available users can accurately compute these metrics without having to recompute them from scratch.

The existing solution computes perplexity score post training by running the model through a validation batch via  `trainer.evaluate()`, this only provides a single score on a single validation batch after all training is done.

### Motivation

Why other solutions to track perplexity during training via `compute_metrics` do not work effectively :

- Using `evaluate` library to load perplexity :  when using Trainer, it [reloads](https://github.com/huggingface/evaluate/blob/main/metrics/perplexity/perplexity.py#L103) the model  from the hub at each evaluation which is inefficient. Furthermore, the implementation is not that flexible, for instance it assumes that `labels` are exactly equal to `input_ids` which is not always the case, for instance I’m running instruction finetuning where the labels are the response part of the `input_ids`.

- Computing perplexity direclty using `EvalPrediction`'s logits and labels to compute the loss again inside of `compute_metrics` function would be inaccurate since we do not have access to `attention_mask` of the validation batches, so we can’t compute proper loss by masking padding tokens.

- Why make loss accessible in `compute_metrics` along with logits and labels :
    - The evaluation loss is computed by default earlier in the evaluation loop, and `compute_metrics` is executed right after.
    - I’m assured that the loss used is accurate and consistent and thus the reported perplexity is reliable. Furthermore, the loss can be used to compute other metrics other than perplexity.
    - I believe this feature would be useful since there are similar request In the community like [this](https://discuss.huggingface.co/t/compute-perplexity-using-compute-metrics-in-sfttrainer/74974) and [this](https://discuss.huggingface.co/t/using-perplexity-as-metric-during-training/42354).

### Your contribution

I would be happy to implement this and get your feedback on whether it is the right approach before opening PR 😊.

### PR DESCRIPTION
**Pull request by Manalelaidouni**
# What does this PR do?
Fixes #32307

This PR includes loss in the `compute_metrics` function via the `include_loss_for_metrics` training argument flag, this is particularly useful for calculating loss dependent metrics, the `EvalPrediction` class now handles optional inputs and losses.

## Before submitting

- [ ]  This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x]  Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),Pull Request section?
- [x]  Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ]  Did you make sure to update the documentation with your changes? Here are the[documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and[here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ]  Did you write any new necessary tests?

## Who can review?

@muellerzr @SunMarc

### CODE REVIEW

File: src/transformers/trainer.py
Comment (SunMarc) [inline src/transformers/trainer.py:4598]: We can further simplify these ! let's remove the if else and call metrics = self.compute_metrics only once.

File: src/transformers/trainer.py
Comment (SunMarc) [inline src/transformers/trainer.py:4647]: same here

File: src/transformers/trainer_utils.py
Comment (SunMarc) [inline src/transformers/trainer_utils.py:166]: let's not change that and create `losses` instead

File: src/transformers/trainer_utils.py
Comment (SunMarc) [inline src/transformers/trainer_utils.py:174]: to update

File: src/transformers/trainer_utils.py
Comment (SunMarc) [inline src/transformers/trainer_utils.py:187]: nice !

File: src/transformers/trainer.py
Comment (SunMarc) [inline src/transformers/trainer.py:4075]: same here

**Review by SunMarc — COMMENTED** [review]
Nice work ! Left a few comments

**Comment by Manalelaidouni** [comment]
Hey @SunMarc, thanks for the great feedback! 
I’ve refactored the code as you suggested and updated the EvalPrediction accordingly

**Review by SunMarc — APPROVED** [review]
Nice work @Manalelaidouni ! Thanks for iterating ! I reran the tests. They are indeed not related to your changes !

File: src/transformers/trainer_utils.py
Comment (amyeroberts) [inline src/transformers/trainer_utils.py:160]: Should be marked as optional in the docstring

File: src/transformers/trainer_utils.py
Comment (amyeroberts) [inline src/transformers/trainer_utils.py:174]: Let's use a different attribute name than `items` here: `items` is typically used for a method which returns an iterable of key, value pairs i.e. the usage is `obj.items()` rather than `obj.items`.

File: src/transformers/trainer_utils.py
Comment (amyeroberts) [inline src/transformers/trainer_utils.py:186]: Much cleaner :)

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Overall looks good - and makes some of the logic a lot cleaner. Main comment is the naming of the `items` attribute. 

I'd like to have another approval from @muellerzr before merge, as this adds another argument to `training_args` and he's recently been working on trying to reduce / refactor this

**Comment by Manalelaidouni** [comment]
Thank you @amyeroberts for the review! 

my bad I had commented yesterday under your review and just realized I had a "pending" label so the comment was only visible on my end.

I’ve already specified the arguments in the docstring as *optional*, are you suggesting that I should add (Optional[np.ndarray], *optional*) or include the phrase defaults to None ?

File: src/transformers/trainer_utils.py
Comment (Manalelaidouni) [reply src/transformers/trainer_utils.py:160]: Thank you @amyeroberts for the review!

my bad I had commented yesterday under your review and just realized I had a "pending" label so the comment was only visible on my end.

I’ve already specified the arguments in the docstring as optional, are you suggesting that I should add (Optional[np.ndarray], optional) or include the phrase defaults to None ?

Comment (amyeroberts) [reply src/transformers/trainer_utils.py:160]: Sorry, my bad, I must have just skimmed and not read properly. All good!

**Review by muellerzr — COMMENTED** [review]
Thanks. I don’t think it makes sense to containerize these args *potentially*. At most I could see us passing in a list in a new `include_for_metrics` which can be “loss”, “input”, etc. WDYT @amyeroberts?

we then deprecate the other include_ for 5.0

**Comment by amyeroberts** [comment]
@muellerzr Sounds good to me!

**Comment by Manalelaidouni** [comment]
@muellerzr great idea! I was actually hesitant to suggest removing  `include_input_for_metrics`  but I agree it would make the API much cleaner; 

Just to make sure we’re on the same page before I move forward with your suggestion, so we’ll set `include_for_metrics` as a `(list[str], *optional*, defaults to [])` — 

To set `include_for_metrics` as an empty list and not `None` (to simplify code 
and not have to check if it’s `None` everytime we use it) we’ll mark the field as `default_factory=list`  — 

We'll also add a deprecation warning for `include_inputs_for_metrics` like this:


```python
if args.include_inputs_for_metrics:
    warnings.warn("Using `include_inputs_for_metrics` is deprecated. Please use `include_for_metrics` instead.",FutureWarning)
    args.include_for_metrics.append("inputs")
```


@muellerzr @amyeroberts does this sound good?

**Comment by amyeroberts** [comment]
@Manalelaidouni Sounds good to me! Only small addition would be that we'd ideally mention when the `include_for_metrics` will be deprecated (v5) and

File: src/transformers/training_args.py
Comment (SunMarc) [inline src/transformers/training_args.py:2069]: We prefer to use logger.warning as users will have more control over their logs 

```suggestion
            logger.warning(
                "Using `include_inputs_for_metrics` is deprecated and will be removed in version 5 of 🤗 Transformers. Please use `include_for_metrics` list argument instead."
            )
```

**Review by SunMarc — APPROVED** [review]
Nice job @Manalelaidouni  ! Thanks for iterating ! I left a small suggestion

**Review by amyeroberts — APPROVED** [review]
Very nice - thanks for adding and iterating! 

+1 on @SunMarc's nit. Otherwise all looks good to me!

**Comment by Manalelaidouni** [comment]
Hi @muellerzr, just wanted to follow up on this PR, would appreciate your review when possible, thanks in advance!

**Comment by OxxoCodes** [comment]
Just bumping this - really useful change, would like to see this PR merged if possible

**Comment by SunMarc** [comment]
Merging it ! Thanks for your work @Manalelaidouni !

**Pull request merged**
